### ISSUE TITLE
How can i convert ImGuiFileDialog for make it usable with CimGui

### ISSUE THREAD

**Issue by aiekick** [cimgui/cimgui#172]
Hello,

i would like to convert my lib [ImGuiFileDialog](https://github.com/aiekick/ImGuiFileDialog) for make it usable with CImGui.
for target more audience with all other bindings :)

but im not familair with C and how CImGui can be generated from ImGui who is as i know for C++.

i imagine i need to  :
 * remove my singleton
 * convert class to struct
 * remove all things from stl (like std::vector, std::map, std::string)

maybe i can furnish a c interface too in the same file

And after when i done all that things, how can i know if it will be converted via your generator to make it usable to CImgui.

is there more things than i need to know ?

Thanks for your help :)

**Comment by sonoro1234** [comment]
> but im not familair with C and how CImGui can be generated from ImGui who is as i know for C++.

Basically cimgui wraps all ImGui functions with C functions that can then be used from a ffi interface capable language. It also provides C structure definitions that are compatible with the library.

> maybe i can furnish a c interface too in the same file

If you do that you wont need cimgui generator at all as the work will be done already.

> And after when i done all that things, how can i know if it will be converted via your generator to make it usable to CImgui.

I could give it a try


> is there more things than i need to know ?

No, that I know now.

The purpose of cimgui is to allow automatically making a C interface for a C++ library which is continuosly developing. Only the header files are providing the info for that. It seems that ImGuiFileDialog would need a lot of changes for being prepared for cimgui, so that an alternative would be to roll your own C API interface.

**Comment by aiekick** [comment]
ok so i can keep my c++ stuff like stl structure internally (like std::string) ?

by ex keep my cpp code 
```cpp
void OpenDialog(const std::string& vKey, const char* vName, const char* vFilters,
			const std::string& vPath, const std::string& vDefaultFileName,
			const std::function<void(std::string, UserDatas, bool*)>& vOptionsPane, const float&  vOptionsPaneWidth = 250.0f,
			const int& vCountSelectionMax = 1, UserDatas vUserDatas = nullptr, ImGuiFileDialogFlags flags = 0);
		
```
and give a c interface like that : 

```cpp
typedef void (*IGFDPaneFun)(const char*, void*, bool*);
IMGUIFILEDIALOG_API void OpenDialog(const char* vKey, const char* vName, const char* vFilters, const char* vPath, 
	const char* vDefaultFileName, IGFDPaneFun vOptionsPane, const float& vOptionsPaneWidth, 
	const int& vCountSelectionMax, void* vUserDatas, ImGuiFileDialogFlags flags);
```

do you know if i can include the c interface funcs in the same cpp code ?

**Comment by sonoro1234** [comment]
> ```c++
> const float& vOptionsPaneWidth
> ```

C++ references (&) have no meaning in C, you should use a const float or const float *



> do you know if i can include the c interface funcs in the same cpp code ?

Yes, you can as the compiling will be a C++ compilation. Just remember to declare the function as extern "C" as done in https://github.com/cimgui/cimgui/blob/master/cimgui.h#L33 where `CIMGUI_API` is adapted for different compilers

**Comment by aiekick** [comment]
OK thank you very mush.

its exiting haha, never really write this kind of things after 17 years of c++ btw :)

i will create a small branch for that and let you know if you can check when its seem working for me :)

Thanks again

**Comment by aiekick** [comment]
More difficult than i thought at first haha.
i have udated my funcs stamp for make it compatible with a c interface i think

i have keeped in some cases the std::string (like when i return string in params) but i can propose a specific c stamp for that.
i wantd keep theirs funcs easy for c++ at first.

all the others was converted to const char *
i have removed my singleton to for have the way in c to create a ImGuiFileDialog context for construction/destruction
so i think he will work.

btw i have many issue for let it work with ImGui structs en enum. im sure i not do that correctly.

Question, maybe the easiest way is to ceate a shared lib of ImGuiFileDialog.
but if do that, i need to use cimgui ? because in my funcs im using some times ImGui types
or maybe just imgui ?
at this moment, if my imgui version used in this dll is different of the version used in cimgui by a guy, can it be an issue ?

thanks for your help

**Comment by sonoro1234** [comment]
> at this moment, if my imgui version used in this dll is different of the version used in cimgui by a guy, can it be an issue ?

My prefered option is to compile all cimgui and optional widgets in the same dll to avoid problems.

**Comment by aiekick** [comment]
so for the c interface i need to include cimgui.h or just imgui.h ?

**Comment by sonoro1234** [comment]
> so for the c interface i need to include cimgui.h or just imgui.h ?

If you are creating your own C interface you dont need to include cimgui.h at all
Do you have a link for the work done?

**Comment by aiekick** [comment]
yes just finished. it seems working for me.

[ImGuiFileDialog C Interface](https://github.com/aiekick/ImGuiFileDialog/tree/c_interface)

There in the main.cpp who show both api c/cpp behavior. 
And the Directory ImGuiFileDialog who contain cpp/h source for ImGuiFIleDialog.

No sure if its well implemented, let me know what i can improve :)

thanks for your help

**Comment by sonoro1234** [comment]
I could build ImGuiFileDialog executable correctly.
I could not compile ImGuiFileDialog.cpp file with the rest of LuaJIT-ImGui files because:

```
[  3%] Building CXX object CMakeFiles/cimgui_base.dir/ImGuiFileDialog/ImGuiFileDialog/ImGuiFileDialog.cpp.obj
C:\LuaGL\gitsources\anima\LuaJIT-ImGui\ImGuiFileDialog\ImGuiFileDialog\ImGuiFileDialog.cpp: In function 'std::vector<std::__cxx11::basic_string<char>
> IGFD::GetDrivesList()':
C:\LuaGL\gitsources\anima\LuaJIT-ImGui\ImGuiFileDialog\ImGuiFileDialog\ImGuiFileDialog.cpp:280:3: error: 'DWORD' was not declared in this scope
   DWORD mydrives = 2048;
   ^~~~~
C:\LuaGL\gitsources\anima\LuaJIT-ImGui\ImGuiFileDialog\ImGuiFileDialog\ImGuiFileDialog.cpp:283:8: error: expected ';' before 'countChars'
   DWORD countChars = mini(GetLogicalDriveStringsA(mydrives, lpBuffer), 2047);
        ^~~~~~~~~~~
        ;
C:\LuaGL\gitsources\anima\LuaJIT-ImGui\ImGuiFileDialog\ImGuiFileDialog\ImGuiFileDialog.cpp:285:7: error: 'countChars' was not declared in this scope
   if (countChars > 0)
       ^~~~~~~~~~
C:\LuaGL\gitsources\anima\LuaJIT-ImGui\ImGuiFileDialog\ImGuiFileDialog\ImGuiFileDialog.cpp:285:7: note: suggested alternative: 'fputchar'
   if (countChars > 0)
       ^~~~~~~~~~
       fputchar
C:\LuaGL\gitsources\anima\LuaJIT-ImGui\ImGuiFileDialog\ImGuiFileDialog\ImGuiFileDialog.cpp: In function 'bool IGFD::CreateDirectoryIfNotExist(const st
ring&)':
C:\LuaGL\gitsources\anima\LuaJIT-ImGui\ImGuiFileDialog\ImGuiFileDialog\ImGuiFileDialog.cpp:325:5: error: 'CreateDirectoryA' was not declared in this s
cope
     CreateDirectoryA(name.c_str(), nullptr);
     ^~~~~~~~~~~~~~~~
C:\LuaGL\gitsources\anima\LuaJIT-ImGui\ImGuiFileDialog\ImGuiFileDialog\ImGuiFileDialog.cpp:325:5: note: suggested alternative: 'CreateDirectoryIfNotEx
ist'
     CreateDirectoryA(name.c_str(), nullptr);
     ^~~~~~~~~~~~~~~~
     CreateDirectoryIfNotExist
C:\LuaGL\gitsources\anima\LuaJIT-ImGui\ImGuiFileDialog\ImGuiFileDialog\ImGuiFileDialog.cpp: In member function 'void IGFD::FileDialog::ScanDir(const s
tring&)':
C:\LuaGL\gitsources\anima\LuaJIT-ImGui\ImGuiFileDialog\ImGuiFileDialog\ImGuiFileDialog.cpp:1927:8: error: 'scandir' was not declared in this scope
    n = scandir(path.c_str(), &files, nullptr, alphaSort);
        ^~~~~~~
C:\LuaGL\gitsources\anima\LuaJIT-ImGui\ImGuiFileDialog\ImGuiFileDialog\ImGuiFileDialog.cpp:1927:8: note: suggested alternative: 'ScanDir'
    n = scandir(path.c_str(), &files, nullptr, alphaSort);
        ^~~~~~~
        ScanDir
C:\LuaGL\gitsources\anima\LuaJIT-ImGui\ImGuiFileDialog\ImGuiFileDialog\ImGuiFileDialog.cpp:1945:20: error: 'struct dirent' has no member named 'd_type
'; did you mean 'd_name'?
       switch (ent->d_type)
                    ^~~~~~
                    d_name
C:\LuaGL\gitsources\anima\LuaJIT-ImGui\ImGuiFileDialog\ImGuiFileDialog\ImGuiFileDialog.cpp:1947:12: error: 'DT_REG' was not declared in this scope
       case DT_REG:
            ^~~~~~
C:\LuaGL\gitsources\anima\LuaJIT-ImGui\ImGuiFileDialog\ImGuiFileDialog\ImGuiFileDialog.cpp:1949:12: error: 'DT_DIR' was not declared in this scope
       case DT_DIR:
            ^~~~~~
C:\LuaGL\gitsources\anima\LuaJIT-ImGui\ImGuiFileDialog\ImGuiFileDialog\ImGuiFileDialog.cpp:1949:12: note: suggested alternative: 'DIR'
       case DT_DIR:
            ^~~~~~
            DIR
C:\LuaGL\gitsources\anima\LuaJIT-ImGui\ImGuiFileDialog\ImGuiFileDialog\ImGuiFileDialog.cpp:1951:12: error: 'DT_LNK' was not declared in this scope
       case DT_LNK:
            ^~~~~~
C:\LuaGL\gitsources\anima\LuaJIT-ImGui\ImGuiFileDialog\ImGuiFileDialog\ImGuiFileDialog.cpp:1951:12: note: suggested alternative: 'M_LN2'
       case DT_LNK:
            ^~~~~~
            M_LN2
C:\LuaGL\gitsources\anima\LuaJIT-ImGui\ImGuiFileDialog\ImGuiFileDialog\ImGuiFileDialog.cpp: In member function 'void IGFD::FileDialog::SetCurrentDir(c
onst string&)':
C:\LuaGL\gitsources\anima\LuaJIT-ImGui\ImGuiFileDialog\ImGuiFileDialog\ImGuiFileDialog.cpp:2014:21: error: 'GetFullPathNameA' was not declared in this
 scope
    size_t numchar = GetFullPathNameA(path.c_str(), PATH_MAX - 1, real_path, nullptr);
                     ^~~~~~~~~~~~~~~~
C:\LuaGL\gitsources\anima\LuaJIT-ImGui\ImGuiFileDialog\ImGuiFileDialog\ImGuiFileDialog.cpp:2014:21: note: suggested alternative: 'GetFilePathName'
    size_t numchar = GetFullPathNameA(path.c_str(), PATH_MAX - 1, real_path, nullptr);
                     ^~~~~~~~~~~~~~~~
                     GetFilePathName
mingw32-make[2]: *** [CMakeFiles\cimgui_base.dir\build.make:315: CMakeFiles/cimgui_base.dir/ImGuiFileDialog/ImGuiFileDialog/ImGuiFileDialog.cpp.obj] E
```

**Comment by aiekick** [comment]
ok thanks i will check with your LUA-jit too

**Comment by aiekick** [comment]
on what system are you ?
just a things, on windows you need dirent.h. (the DT_DIR, DT_FILE, etc..). on unix dirent is system
GetFilePathName and so on are windows only

**Comment by sonoro1234** [comment]
> on what system are you ?

I am compiling in windows. Just added ImGuiFileDialog.cpp to source files of LuaJIT-ImGui compilation.

**Comment by aiekick** [comment]
i just done that too.
just added dirent directory in ImGuiFIleDialog directory
add added that in the CMakeLists.txt
```c
if(WIN32)
    include_directories(${CMAKE_CURRENT_SOURCE_DIR}/ImGuiFileDialog/dirent/include)
endif()
```
and the compilation was ok in VS2019.

**Comment by aiekick** [comment]
> I am compiling in windows. Just added ImGuiFileDialog.cpp to source files of LuaJIT-ImGui compilation. (there is no error in dirent.h inclusion)
> 
DT_REG, DT_DIR_DT_LNK are declared by dirent only

> By the way, I have https://github.com/sonoro1234/luafilesystem/blob/unicode/lfs_ffi.lua where filesystem functions for LuaJIT are implemented emulating Lua luafilesystem module, It is done just with standard system functions for windows (with unicode), linux and macos.

and you think its an issue ?

**Comment by aiekick** [comment]
i will make it work with your cimgui repo. i will create a dir in parallel of imgui in cimgui.
i know now, i need to include dirent dir in ImGuiFileDialo.
i seen also some missing struct keywords in fucn call and other errors
the goal is just to have the possibility to use it with C if needed. 
i received many queries by the apst for that :)

**Comment by sonoro1234** [comment]
> DT_REG, DT_DIR_DT_LNK are declared by dirent only, and theirs is error here

Yes, there is a dirent error but not for inclusion as the compiler would error in `#include <dirent.h>`. The reason is that mingw-w64 has dirent.h already but with DT_REG and related not defined in it.

But if I add cmake lines for include_directories to your dirent.h file compilation succeds.

> and you think its an issue ?

Not an issue, but  I think that lfs_ffi.lua is more minimalist (which I always prefer)

**Comment by aiekick** [comment]
ok thanks for your help :)

**Comment by sonoro1234** [comment]
I could also did a quick and dirty test with the compiled dll that seems to work

```Lua
local igwin = require"imgui.window"
--local win = igwin:SDL(800,400, "widgets")
local win = igwin:GLFW(800,400, "widgets")
local ig = win.ig

local ffi = require"ffi"

ffi.cdef[[
typedef struct ImGuiFileDialog ImGuiFileDialog;
ImGuiFileDialog* IGFD_Create(void);
typedef void (*IGFD_PaneFun)(const char*, void*, bool*);
typedef int ImGuiFileDialogFlags;
void IGFD_OpenDialog(ImGuiFileDialog* vContext,
		const char* vKey, const char* vName, const char* vFilters, const char* vPath,
		const char* vDefaultFileName, IGFD_PaneFun vOptionsPane, const float vOptionsPaneWidth,
		const int vCountSelectionMax, void* vUserDatas, ImGuiFileDialogFlags flags);
enum ImGuiFileDialogFlags_
{
	ImGuiFileDialogFlags_None = 0,
	ImGuiFileDialogFlags_ConfirmOverwrite = 1 << 0,
};
bool IGFD_DisplayDialog(ImGuiFileDialog* vContext, 
		const char* vKey, ImGuiWindowFlags vFlags, ImVec2 vMinSize, ImVec2 vMaxSize);
void IGFD_CloseDialog(ImGuiFileDialog* vContext);
bool IGFD_IsOk(ImGuiFileDialog* vContext);
struct IGFD_String
	{
		char* buffer;	// 0
		size_t size;	// 0U
	};
struct IGFD_String IGFD_GetFilePathName(ImGuiFileDialog* vContext);
]]


local canValidateDialog = ffi.new("bool[1]",false);
local cfileDialog = ig.lib.IGFD_Create()
local function InfosPane(vFilter, vUserDatas, vCantContinue) -- if vCantContinue is false, the user cant validate the dialog
	ig.TextColored(ig.ImVec4(0, 1, 1, 1), "Infos Pane");
	
	ig.Text("Selected Filter : %s", vFilter);

	local userDatas = vUserDatas;
	if (userDatas) then
        ig.Text("User Datas : %s", userDatas);
	end

	ig.Checkbox("if not checked you cant validate the dialog", canValidateDialog);

	if (vCantContinue) then
	    vCantContinue[0] = canValidateDialog[0];
	end
end

function win:draw(ig)

	if ig.Begin("widgets",nil, ig.lib.ImGuiWindowFlags_AlwaysAutoResize) then
		if ig.Button("FileDialog") then
			ig.lib.IGFD_OpenDialog(cfileDialog, "ChooseFileDlgKey",
							--ICON_IGFD_SAVE " Choose a File", filters,
							 " Choose a File", filters,
							".", "", InfosPane, 350, 1, ffi.cast("void*","SaveFile"), ig.lib.ImGuiFileDialogFlags_ConfirmOverwrite);
		end
		local minSize = ig.ImVec2(0, 0);
		local maxSize = ig.ImVec2(ig.FLT_MAX, ig.FLT_MAX);
		if (ig.lib.IGFD_DisplayDialog(cfileDialog, "ChooseFileDlgKey",
					ig.lib.ImGuiWindowFlags_NoCollapse, minSize, maxSize)) then
			if (ig.lib.IGFD_IsOk(cfileDialog)) then
				local sss = ig.lib.IGFD_GetFilePathName(cfileDialog);
				print(sss, ffi.string(sss.buffer))
			end
			ig.lib.IGFD_CloseDialog(cfileDialog);
		end
	end
	ig.End()
end

win:start()
```

**Comment by aiekick** [comment]
ok nice, yes please :)

**Comment by sonoro1234** [comment]
> or its like an export for make them available in Lua FFI ?

That's it

**Comment by aiekick** [comment]
It weird i have modified your project cimgui for include ImGuiFIleDialog
cimgui_dll compil fine,
im well in C (not cpp)
see (the cpp defined are disabled):
![devenv_MfGaXlrElp](https://user-images.githubusercontent.com/1434736/104234983-8fb63080-5454-11eb-92e5-a1053f6f26d1.png)

but after i have some not found symbol.. dont know what can be the cause
![devenv_GzTU6FWBU9](https://user-images.githubusercontent.com/1434736/104234985-90e75d80-5454-11eb-9c2d-c5633158aae0.png)

**Comment by aiekick** [comment]
i have created that [here](https://github.com/aiekick/cimgui)

**Comment by aiekick** [comment]
ok just foudn the issue. the fucntion was not exported in the dll.
taken your define for the declexport. 
since it working now i will modify my func stamp for make it easy at most ofr c and c++ :)

**Comment by aiekick** [comment]
i was return from my funcs a struct named IGFD_String who was containing the char buffer and his size.
but i thivck i can just return the char buffer. size is not needed for delete array of char.

or you think it can be usefull in some case ? my goal is to simplify at most

**Comment by sonoro1234** [comment]
Better char* only

**Comment by aiekick** [comment]
i have finalized my api here : https://github.com/aiekick/ImGuiFileDialog
just waiting the release on ImGui 1.80 (for the moment the table api change every day's :) )
thanks for your great support :)

**Comment by aiekick** [comment]
its released now : [ImGuiFileDialog v0.5.5](https://github.com/aiekick/ImGuiFileDialog/releases/tag/v0.5.5)

:)
